imap/jmap_calendar.c

```diff
@@ -1567,47 +1566,43 @@ static int _calendarevent_getblob_cb(const char *mailbox __attribute__((unused))
     if (!buf_len(value)) return 0;
 
     struct calendarevent_getblob_rock *rock = vrock;
-    jmap_req_t *req = rock->req;
     struct buf *buf = rock->buf;
 
     /* Parse the value and fetch the patch */
     struct dlist *dl;
     const char *vpatchstr = NULL;
     dlist_parsemap(&dl, 1, 0, buf_base(value), buf_len(value));
     dlist_getatom(dl, "VPATCH", &vpatchstr);
-    if (vpatchstr) buf_setcstr(buf, vpatchstr);
-    dlist_free(&dl);
-    if (!buf_len(buf)) return 0;
-
-    /* Write VPATCH blob */
-    char *part_headers = NULL;
-    if (userid) part_headers = strconcat("X-UserId: ", userid, "\r\n", NULL);
-    req->txn->resp_body.type = "text/calendar; component=VPATCH";
-    req->txn->resp_body.len = buf_len(buf);
-    write_multipart_body(0, req->txn, buf_base(buf), buf_len(buf), part_headers);
-    free(part_headers);
+    if (vpatchstr) {
+        /* Write VPATCH blob */
+        buf_printf(buf, "\r\n--%s\r\n", rock->boundary);
+        buf_appendcstr(buf, "Content-Type: text/calendar; component=VPATCH\r\n");
+        buf_printf(buf, "Content-Length: %zu\r\n", strlen(vpatchstr));
+        if (userid) buf_printf(buf, "X-UserId: %s\r\n", userid);
+        buf_appendcstr(buf, "\r\n");
+        buf_appendcstr(buf, vpatchstr);
+    }
 
+    dlist_free(&dl);
     return 0;
 }
 
-static int jmap_calendarevent_getblob(jmap_req_t *req,
-                                      const char *blobid,
-                                      const char *accept_mime)
+static int jmap_calendarevent_getblob(jmap_req_t *req, jmap_getblob_context_t *ctx)
 {
     struct caldav_db *db = NULL;
     struct caldav_data *cdata = NULL;
     struct mailbox *mailbox = NULL;
     icalcomponent *ical = NULL;
+    static struct buf cbuf = BUF_INITIALIZER;
```

Comment:
Is using a static buffer really necessary here? We are mallocing all over the place in the blob code, so if static cbuf is just an optimization, I don't think it is warranted (and could cause future subtle bugs).